File: Data Structures/Queue/Java/Linear_Queue_Program.java
import java.util.Scanner;

class Queue{
    int size, front, rear, cap, deq_element;
    int[] queue;
    public Queue(int capacity){
        this.cap = capacity;
        front = rear = -1;
        size = 0;
        queue = new int[capacity];
    }
    public boolean isFull(){
        if (size == cap){
            return true;
        }else{
            return false;
        }
    }
    public boolean isEmpty(){
        if (size == 0){
            return true;
        }else{
            return false;
        }
    }
    public void enqueue(int value){
        if (isFull() || rear == cap - 1){
            System.out.println("Queue is full");
        }else{
            rear++;
            queue[rear] = value;
            size++;
            System.out.println(value + " enqueued in the Queue.\n");
            if (front == -1){
                front = 0;
            }
        }
    }
    public int dequeue(){
        if (isEmpty()){
            return -1;
        }else{
            deq_element = queue[front];
            size--;
            front++;
            if (size == 0){
                front = rear = -1;
            }
            return deq_element;
        }
    }
    public int peek(){
        if (isEmpty()){
            return -1;
        } else{
            return queue[front];
        }
    }
    public void print(){
        if (isEmpty()){
            System.out.println("Queue is Empty\n");
        }else{
            System.out.print("Queue elements are: ");
            for(int i = front; i <= rear; i++){
                System.out.print(queue[i] + " ");
            }
            System.out.println("\n");
        }
    }
}
public class Linear_Queue_Program{
    public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);
        int capacity, opr, value, choice;
        System.out.print("Enter the capacity of the Queue: ");
        capacity = scanner.nextInt();
        System.out.println();
        System.out.print("Enter the number of operations: ");
        opr = scanner.nextInt();
        System.out.println();
        Queue obj = new Queue(capacity);
        while(opr-->0){
            System.out.println("Enter the choice:\n\n1. Enqueue\n2. Dequeue\n3. Peek\n4. Print\n");
            choice = scanner.nextInt();
            switch(choice){
                case 1: 
                    System.out.print("\nEnter the element you want to enqueue: ");
                    value = scanner.nextInt();
                    System.out.println();
                    obj.enqueue(value);
                    break;
                case 2: 
                    value = obj.dequeue();
                    if (value == -1 && obj.isEmpty() == true){
                        System.out.println("Queue is empty\n");
                    }else{
                        System.out.println("\n" + value + " is dequeued from the Queue.\n");
                    }
                    break;
                case 3:
                    value = obj.peek();
                    if (value == -1 && obj.isEmpty() == true){
                        System.out.println("Queue is empty\n");
                    }else{
                        System.out.println("\n" + value + " is the front element of the Queue.\n");
                    }
                    break;
                case 4:
                    obj.print();
                    break;
                default:
                    System.out.println();
                    System.out.println("Invalid choice!\n");
                    opr++;
                    break;
            }
        }
    }
}

File: Data Structures/Queue/C++/Linear_Queue_Program.cpp
#include<iostream>
using namespace std;

class Queue{
    public:
        int *queue, size, front, rear, cap;
        Queue(int capacity){
            front = rear = -1;
            size = 0;
            queue = new int[capacity];
            cap = capacity;
        }

        bool isFull(){
            if (size == cap){
                return true;
            }
            else{
                return false;
            }
        }
        bool isEmpty(){
            if (size == 0){
                return true;
            }
            else{
                return false;
            }
        }
        void enqueue(int val){
            if(isFull()){
                cout << "Queue is Full!" << endl << endl;
            }
            else{
                rear++;
                queue[rear] = val;
                size++;
                cout << val << " enqueued in the Queue." << endl << endl;
                if (front == -1){
                    front = 0;
                }
            }
        }
        void dequeue(){
            if(isEmpty()){
                cout << "Queue is Empty!" << endl << endl;
            }
            else{
                cout << queue[front] << " is dequeued from the Queue" << endl << endl;
                front++;
                size--;
            }
        }
        void peek(){
            if (isEmpty()){
                cout << "Queue is Empty!" << endl << endl;
            }
            else{
                cout << "The front element of the Queue is: " << queue[front] << endl << endl;
            }
            
        }
        void print(){
            if (isEmpty()){
                cout << "Queue is Empty!" << endl << endl;
            }
            else{
                cout << "Queue elements are:" << " ";
                for(int i = front; i <= rear; i++){
                    cout << queue[i] << " ";
                }
                cout << endl << endl;
            }
        }
        ~Queue(){
            delete[] queue;
        }
};

int main(){
    int capacity;
    cout << "Enter the capacity of the Queue:" << " ";
    cin >> capacity;
    cout << endl;
    Queue obj(capacity);
    int opr;
    cout << "Enter the number of operations:" << " ";
    cin >> opr;
    cout << endl;
    int choice, element;
    while (opr != 0){
        cout << "Enter the choice:\n\n1. Enqueue\n2. Dequeue\n3. Peek\n4. Print" << endl << endl;
        cin >> choice;
        opr--;
        switch(choice){
            case 1: 
                cout << "\nEnter the element you want to enqueue:" << " ";
                cin >> element;
                cout << endl;
                obj.enqueue(element);
                break;
            case 2:
                cout << endl;
                obj.dequeue();
                break;
            case 3:
                cout << endl;
                obj.peek();
                break;
            case 4:
                cout << endl;
                obj.print();
                break;
            default:
                cout << endl;
                cout << "Invalid choice!" << endl << endl;
                opr++;
                break;
        }
    }
    

    return 0;
}
